content/advent-2015/reducing-boilerplate-with-go-generate.md

```diff
@@ -0,0 +1,294 @@
+
+* `go/build`: gathers information about a go package. Given a package
+  name, it'll return information such as what's the directory containing
+the source code for the package, what are the code and test files in the
+directory, what other packages it's dependent on, etc.
+* `go/scanner` and `go/parser`: read source code and parse it to
+  generate an [Abstract Syntax Tree][ast] (AST).
+* `go/ast`: declares the types used to represent the AST and includes
+  some methods to help walking and modifying the tree.
+* `go/types`: declares the data types and implements the algorithms used
+  for type-checking Go packages. While `go/ast` contains the raw tree,
+  this package does all the heavy lifting to process the AST so you can
+  get information about types directly.
+
+## Generating code
+
+When generating code, most projecst just rely on the good old
+`text/template` to generate the code.
+
+I recommend starting generated files with a comment clarifying the code
+is automatically generated, which program generated it and mentioning it
+should not be editign by hand.
+
+```go
+/*
+* CODE GENERATED AUTOMATICALLY WITH github.com/ernesto-jimenez/gogen/unmarshalmap
+* THIS FILE SHOULD NOT BE EDITED BY HAND
+*/
+```
+
+We can also use the `go/format` package to format our code before
+writing it. This package contains the logic used by `go fmt`.
+
+## go generate
+
+Once we start writing tools that generate source code for our programs
+two questions appear quickly: at what point in our development process do we
+generate the code? how do we keep the generated code up to date?
+
+
+Since 1.4, the go tool comes with the `generate` command. It allows us
+to run the tools we use for code generation with the go tool itself. We
+can specify what commands need to be run using special comments within
+our source code and `go generate` will do the work for us.
+
+We just need to add a comment with the following format:
+
+```
+//go:generate shell command
+```
+
+Once you have that, `go generate` will automatically call `command`
+whenever it's run.
+
+There's two points that are important to remember:
+
+* `go generate` is meant to be run by the developer authoring the
+  program or package. It's never called automatically by `go get`
+* You need to have all the tools invoked by `go generate` already
+  installed and setup in your system. Make sure you document what tools
+  you are going to use and where those tools can be downloaded.
+
+Also, if your code generation tool is within the same repository, I
+would recommend calling `go run` from `go:generate`. That way, you can run
+`generate` without building and installing the tool manually each time
+you change the tool.
+
+# How do you start building your own tools?
+
+The stdlib packages to parse and generate code are great, but their
+documentation is huge and it can be quite daunting to make a sense of
+how to use the packages just from the docs.
```

Comment:
documentation is huge and making sense of how to use the packages just from the docs can be quite daunting.